### PR TITLE
EVG-17142 GitHub conditional requests

### PR DESCRIPTION
**Pull request by ybrill**
[EVG-17142](https://jira.mongodb.org/browse/EVG-17142)

### Description
Use [conditional requests](https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#conditional-requests) for calls to the GitHub API. They don't count against the rate limit if we don't have to get the actual response. 

The SDK suggests using [httpcache](https://github.com/gregjones/httpcache) for the cache. If we'd use the default, in-memory, cache it would be specific to each app server. Instead, this PR uses a collection in the db so it can be shared between app servers. My plan is to make the collection with a tight TTL index on when a key was last inserted. If the cache proves useful it will be straightforward to switch it out for a redis cluster in Kanopy.

The cache implementation swallows errors (with logging, but still). This is okay because worst case the cache will have outdated stuff in it and we'll just make more calls to GitHub when we see it's no longer valid.

### Testing
In staging, the first request made a call to GitHub, but subsequent requests did not. A query like [this one](https://ui.honeycomb.io/mongodb-4b/environments/staging/datasets/evergreen/result/nL1dyyQFByn) will tell us how often the response is coming from the cache.

### CODE REVIEW

**Comment by ybrill** [comment]
cc @bynn there were many merge conflicts with #6662. I think I've addressed them all, but if you want to make sure I've captured your intent.

**Comment by ybrill** [comment]
The build will pass (hopefully) once https://github.com/evergreen-ci/utility/pull/53 merges.

File: thirdparty/github.go
Comment (malikchaya2) [inline thirdparty/github.go:263]: nit: do we maybe want to add a comment saying that the client will use caching?

File: thirdparty/github.go
Comment (malikchaya2) [inline thirdparty/github.go:370]: @bynn this reworks the github client code so you may want to review it to see how it impacts EVG-19966.

File: thirdparty/github.go
Comment (malikchaya2) [inline thirdparty/github.go:798]: I see the old code looping through a list of tags while this one seems to parse only one tag. However, the url seems to be the same. Where is the discrepancy from?

File: thirdparty/github.go
Comment (malikchaya2) [inline thirdparty/github.go:943]: Will this pull the value from the cache and give us inaccurate results?

File: thirdparty/github.go
Comment (malikchaya2) [inline thirdparty/github.go:1061]: Do we maybe not want to use the cached value here since using an outdated value here can have bigger repercussions?

File: thirdparty/github.go
Comment (malikchaya2) [inline thirdparty/github.go:1335]: why are we using the legacy token here instead of using it only as a fallback?

File: thirdparty/github.go
Comment (malikchaya2) [inline thirdparty/github.go:1377]: same qeustion as above

File: thirdparty/github.go
Comment (ablack12) [inline thirdparty/github.go:115]: "so we don't rely on a response to not ask" -- this sentence kind of confuses me 😅  why don't we want to rely on max age? why can't we set cachingTransport here directly as opposed to having this extra variable maxAgeTransport?

File: thirdparty/github.go
Comment (ablack12) [inline thirdparty/github.go:286]: can we do this after getting config?

File: thirdparty/github.go
Comment (malikchaya2) [inline thirdparty/github.go:15]: Random comment placement: 
I had two general questions: 
How will the ttl index be set up, aka after what amount of time will the cache expire? We might want to tweak it until we get it right, so it might be a good idea to document somewhere what it's set to and how to change it.

Once we move over to github App, will we keep this caching or remove it?

**Review by malikchaya2 — CHANGES_REQUESTED** [review]

File: thirdparty/github.go
Comment (ablack12) [reply thirdparty/github.go:286]: Oh or could you explain why we do false here and true later?

File: thirdparty/github.go
Comment (ablack12) [inline thirdparty/github.go:341]: what's this bit about?

**Review by ablack12 — CHANGES_REQUESTED** [review]
@bynn can you take a look and verify that this refactor of the installation token looks good to you?

File: thirdparty/github.go
Comment (ybrill) [reply thirdparty/github.go:265]: done

File: thirdparty/github.go
Comment (ybrill) [reply thirdparty/github.go:800]: I'm not sure why we were doing it differently with the app than with the legacy token. To me it seems fraught to list all the tags on the repository since the beginning of time. Also, I think if we wanted to do that we'd need to use the options to get all the pages ([the docs](https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#list-repository-tags) say we'll get a max of 100 at a time).

@bynn was it intentional to have them use different endpoints?

File: thirdparty/github.go
Comment (ybrill) [reply thirdparty/github.go:1063]: discussed above

File: thirdparty/github.go
Comment (ybrill) [reply thirdparty/github.go:15]: 😆 .
I was thinking like a week. Like probably most things won't be useful after that. I mean, it's not really a big deal if we keep things too long unless the storage becomes an issue. And if we kill things too fast the worst case is that we'll ask GH for it again.

File: thirdparty/github.go
Comment (ybrill) [reply thirdparty/github.go:115]: For some routes GitHub sets a max-age in the response to 60 seconds. This would mean that if we request the same thing from GitHub within 60 seconds the cache layer wouldn't even make a conditional request and instead just pull it out of the cache. This seems like a bad idea for our use case where things can change quickly. 

The idea of this extra transport is to override the max-age from the response by telling the cache that we want a max-age of 0. If the cache layer gets a request that sets max-age to something different than the response it will go with the request and make the conditional request according to the request.

File: thirdparty/github.go
Comment (ybrill) [reply thirdparty/github.go:288]: The idea here was to add an attribute to the span for the GitHub call so we could count the ratio of the which token got used by route. `SetAttributes` overwrites any previous attributes with the same key so this set it to false originally and only true if the app token is used.

File: thirdparty/github.go
Comment (ybrill) [reply thirdparty/github.go:343]: This is to let us query the ratio of requests that get their answer from the cache (so they don't count towards the rate limit) to the requests that need to get an original response from GitHub. This will help us to gauge how effective the cache is.

File: thirdparty/github.go
Comment (ybrill) [reply thirdparty/github.go:15]: It seems like there's not much of a downside to using the cache and if it saves requests then why remove it? 
I think the decision will hinge on whether we see it effectively reducing the number of calls that count against the rate limit.

Comment (malikchaya2) [reply thirdparty/github.go:15]: Are we not worried about the data becoming outdated?

File: thirdparty/github.go
Comment (malikchaya2) [reply thirdparty/github.go:1337]: yes, I forgot that getInstallationToken was reconfigured to do that, my bad.

File: thirdparty/github.go
Comment (ablack12) [reply thirdparty/github.go:115]: Thanks for the info! maybe we could reword to "so we don't use Github's max-age, which prevents us from asking if there's been a change if we requested recently" or something like that to remove the double negative?

File: thirdparty/github.go
Comment (ablack12) [inline thirdparty/github.go:302]: we may want to add a log here for EVG-19966

File: thirdparty/github.go
Comment (ablack12) [inline thirdparty/github.go:103]: async discussed adding an ADR to explain briefly why we added this, and what different pieces of it do (the span github cache attribute, the different transporters)

File: model/cache/cache.go
Comment (ablack12) [inline model/cache/cache.go:75]: Do values get deleted from the cache when they're stale or etc?

File: thirdparty/github.go
Comment (malikchaya2) [reply thirdparty/github.go:943]: this is awesome! thanks for explaining it!

File: thirdparty/github.go
Comment (malikchaya2) [reply thirdparty/github.go:15]: That makes sense, thank you!

**Review by malikchaya2 — APPROVED** [review]
LGTM from my end but I think this should wait on a review from Bynn as well.

File: thirdparty/github.go
Comment (bynn) [reply thirdparty/github.go:798]: the "https://api.github.com/repos/%s/%s/git/refs/tags/%s" doesn't seem to exist on the client that we're using. 

It is technically the same route but without the "tags/{tag}" part at the end. It probably makes more sense to call it directly using the route since using the client will make us do extra work

File: thirdparty/github.go
Comment (ybrill) [reply thirdparty/github.go:300]: @bynn and I discussed logging. I pushed f1f9d70 to address his concerns, pending his review.

File: thirdparty/github.go
Comment (ybrill) [reply thirdparty/github.go]: done

File: thirdparty/github.go
Comment (ybrill) [reply thirdparty/github.go:102]: done

File: model/cache/cache.go
Comment (ybrill) [reply model/cache/cache.go:75]: Usually they don't. The TTL index will be deleting most of them

File: thirdparty/github.go
Comment (ybrill) [reply thirdparty/github.go:965]: I'd originally thought about switching to [this route](https://docs.github.com/en/rest/git/refs?apiVersion=2022-11-28#get-a-reference) to get the SHA for the tag (because it's documented and [implemented by the SDK](https://github.com/google/go-github/blob/ae774d0751d40497e72d1ef2e3b98160adffd93e/github/git_refs.go#L50-L68)) but decided to not change unrelated things.

File: docs/decisions/2023-06-21_github_api_cache.md
Comment (ablack12) [inline docs/decisions/2023-06-21_github_api_cache.md:16]: The greater than 16 MB document thing is actually an interesting question, since that could be possible in the future (albeit unlikely for an ungenerated file) -- if the thing doesn't exist in our cache, will we still get it from Github even if it says we should've cached it (or is that a future improvement / thing to investigate)?

File: model/cache/cache.go
Comment (ablack12) [reply model/cache/cache.go:75]: Will we be introducing a TTL for this collection or just for later? I hesitate to not TTL these since things like project yamls are huge and could make this collection pretty big pretty fast.

File: thirdparty/github.go
Comment (ablack12) [inline thirdparty/github.go:113]: duplicate "so we don't" haha

**Review by ablack12 — APPROVED** [review]
Left a couple of comments but overall looks good to me!

File: docs/decisions/2023-06-21_github_api_cache.md
Comment (ybrill) [reply docs/decisions/2023-06-21_github_api_cache.md:16]: yes, if a thing is larger than the limit the Set function will fail and we won't cache it.
This seems okay atm assuming we'll more this to redis if it's working as expected.

File: thirdparty/github.go
Comment (ybrill) [reply thirdparty/github.go]: we _really_ won't
done

File: model/cache/cache.go
Comment (ybrill) [reply model/cache/cache.go:75]: before this merges I'll be creating the collection and making a TTL index on it.

**Review by bynn — APPROVED** [review]
Thanks for waiting! I tested all the changes and I think it should look fine 
There's a couple of errors in staging caused by the fact that the app is only installed on sandbox and not evergreen but that should go away after you merge main

**Pull request merged**
